apps/desktop/src/lib/state/uiState.svelte.ts

```diff
@@ -0,0 +1,147 @@
+import { reactive, type Reactive } from '@gitbutler/shared/storeUtils';
+import {
+	createEntityAdapter,
+	createSlice,
+	type EntityState,
+	type ThunkDispatch,
+	type UnknownAction
+} from '@reduxjs/toolkit';
+
+type DrawerPage = 'commit' | 'pr' | 'br' | undefined;
+
+/**
+ * Stateful properties for the UI, with redux backed fine-grained reactivity.
+ */
+export class UiState {
+	private state = $state<EntityState<UiStateVariable, string>>(uiStateSlice.getInitialState());
+
+	/** Properties scoped to a specific project. */
+	readonly props = this.buildScopedProps({
+		drawer: {
+			page: undefined as DrawerPage
+		},
+		commitGoesHere: {
+			parentId: undefined as string | undefined,
+			branchName: undefined as string | undefined
+		}
+	});
+
+	/** Properties that are globally scoped. */
+	readonly global = this.buildGlobalProps({
+		drawer: {
+			height: 10
+		}
+	});
+
+	constructor(
+		reactiveState: Reactive<typeof this.state>,
+		private dispatch: ThunkDispatch<any, any, UnknownAction>
+	) {
+		$effect(() => {
+			this.state = reactiveState.current;
+		});
+	}
+
+	private getById(id: string, defaultValue: UiStateValue): Reactive<UiStateValue> {
+		const item = $derived(selectById(this.state, id));
+		return reactive(() => (item?.value !== undefined ? item.value : defaultValue));
+	}
+
+	private update(id: string, value: UiStateValue) {
+		this.dispatch(upsertOne({ id, value }));
+	}
+
+	/**
+	 * Generate redux backed properties corresponding to the shape of the
+	 * parameter value, with types corresponding to their default values.
+	 */
+	private buildGlobalProps<T extends DefaultConfig>(param: T): GlobalStore<T> {
+		const props = {} as GlobalStore<DefaultConfig>;
```

Comment:
nit pick: use a declaration constraint instead of a type cast.
```suggestion
		const props: GlobalStore<DefaultConfig> = {};
```